src/drivers/intel/ssp/ssp.c

```diff
@@ -766,19 +780,80 @@ static int ssp_get_hw_params(struct dai *dai,
 	return 0;
 }
 
-/* SSP hw params */
+/*
+ * enabled clock MCLK/BCLK to support codec that need MCLK/BCLK
+ * This is called before trigger(COMP_TRIGGER_START)
+ */
 static int ssp_hw_params(struct dai *dai,
 			 struct sof_ipc_stream_params  *params)
 {
-	dai_info(dai, "ssp_hw_params()");
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+	int ret;
 
+	dai_info(dai, "ssp_hw_params() params.clks_control 0x%x", ssp->params.clks_control);
+
+	spin_lock(&dai->lock);
+
+	if ((ssp->params.clks_control &
+		 SOF_DAI_INTEL_SSP_CLKCTRL_MCLK_ES)) {
+		ret = ssp_set_mclk(dai);
+		if (ret < 0)
+			goto out;
+	}
+
+	if ((ssp->params.clks_control &
+		 SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES)) {
+		ret = ssp_set_bclk(dai);
+		if (ret < 0)
+			goto out;
+		/* enable port */
+		ssp_update_bits(dai, SSCR0, SSCR0_SSE, SSCR0_SSE);
+	}
+out:
+
+	spin_unlock(&dai->lock);
```

Comment:
Im getting super confused here. We do the same things in ssp_hw_params() and ssp_pre_start(). Why do we need to do these twice?

Reply:
> Im getting super confused here. We do the same things in ssp_hw_params() and ssp_pre_start(). Why do we need to do these twice?

They are not called twice. We do them in ssp_hw_params() if `SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES` is set, and do them in ssp_pre_start() if `SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES` is not set. And same thing as MCLK.